### PR TITLE
Enhancements and New Feature Implementation: Fix rag and new tools

### PR DESCRIPTION
**Pull request by moarshy**
This PR encompasses the following updates:

1. **Bug Fixes for Prediction-Request RAG Tool**: We've identified and resolved several issues affecting the prediction-request RAG tool, enhancing its stability and performance. These fixes address bugs that were impacting the tool's accuracy.

2. **Conversion of Market Reasoning Tool**: The market reasoning tool has been doing very well on the benchmark. We have converted that into a tool for the trader under `prediciton_request_reasoning`.

3. **New Tool: Chain of Thought Document Analysis**: We have had some success with a new tool implementing COT on document/article retrieved from single URL. If a prediction is not reached, we will use the next url. This tool also performs well in the benchmark. This is under `prediction_url_cot`. On benchmark it scored `0.68`.

All can be found under napthaai.

Please review the changes and let us know if there is anything.

### CODE REVIEW

File: packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py [hunk 1]
```diff
@@ -572,7 +597,6 @@ def run(**kwargs) -> Tuple[str, Optional[Dict[str, Any]]]:
             raise ValueError(f"Tool {tool} is not supported.")
 
         engine = TOOL_TO_ENGINE[tool]
-
         additional_information, counter_callback = fetch_additional_information(
```
Comment (jmoreira-valory) [inline packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py:600]: Shouldn't pass temperature and max_tokens here as well?

**Review by Adamantios — COMMENTED** [review]
LGTM, can you link the benchmark's results for `prediction_url_cot` in the description for reference?

File: packages/napthaai/customs/prediction_url_cot/prediction_url_cot.py [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+            max_tokens=max_tokens,
+            temperature=temperature,
+        )
+        
+        if answer.prediction in ["yes", "no"]:
+            return answer, counter_callback, prediction_prompt
+        
+    return Results(p_yes=0.5, p_no=0.5, confidence=0.5, info_utility=0.5, prediction="I don't know"), counter_callback, prediction_prompt
+
+def run(**kwargs) -> Tuple[str, Optional[Dict[str, Any]]]:
```
Comment (jmoreira-valory) [inline packages/napthaai/customs/prediction_url_cot/prediction_url_cot.py:516]: Please fix the interface to match the requirements in https://github.com/valory-xyz/mech/blob/main/README.md#build-your-own

File: packages/napthaai/customs/prediction_url_cot/prediction_url_cot.py [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+        pairs = str(results).split()
+        result_dict = {}
+        for pair in pairs:
+            key, value = pair.split("=")
+            if key != "prediction":
+                result_dict[key] = float(value)
+
+        results = result_dict
+        results = json.dumps(results)
+        return results, prediction_prompt, None, counter_callback
```
Comment (jmoreira-valory) [inline packages/napthaai/customs/prediction_url_cot/prediction_url_cot.py:574]: Fix to match the interface.

File: packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+    pattern = r'\"(.*?)\"'
+    try:
+        question = re.findall(pattern, prompt)[0]
+    except Exception as e:
+        question = prompt
+
+    return question
+
+
+def run(**kwargs) -> Tuple[Optional[str], Optional[Dict[str, Any]], Any]:
```
Comment (jmoreira-valory) [inline packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py:682]: Fix the interface.

File: packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py [hunk 5]
```diff
@@ -0,0 +1,794 @@
...
+        if counter_callback is not None:
+            counter_callback(
+                input_tokens=response_reasoning.usage.prompt_tokens
+                + response.usage.prompt_tokens,
+                output_tokens=response_reasoning.usage.completion_tokens
+                + response.usage.completion_tokens,
+                model=engine,
+                token_counter=count_tokens,
+            )
+        return results, reasoning_prompt + "////" + prediction_prompt, None, counter_callback
```
Comment (jmoreira-valory) [inline packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py:794]: Fix the interface.

**Review by jmoreira-valory — CHANGES_REQUESTED** [review]
Please, fix the interface of run methods according to https://github.com/valory-xyz/mech/blob/main/README.md#build-your-own, both at definition, and at any `return` statement within. 

Note that second output value is an `Optional[Dict[str, Any]]`, and most times it is wrongly output a string.

A few instances have been marked in the review, but please, verify to cover all of them.

File: packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py [hunk 6]
```diff
@@ -65,13 +65,13 @@ def __exit__(self, exc_type, exc_value, traceback) -> None:
...
 MAX_TOKENS = {
-    "gpt-3.5-turbo": 4096,
+    "gpt-3.5-turbo-0125": 16385,
     "gpt-4": 8192,
 }
 ALLOWED_TOOLS = [
     "prediction-request-rag",
 ]
-TOOL_TO_ENGINE = {tool: "gpt-3.5-turbo" for tool in ALLOWED_TOOLS}
+TOOL_TO_ENGINE = {tool: "gpt-3.5-turbo-0125" for tool in ALLOWED_TOOLS}
```
Comment (0xArdi) [inline packages/napthaai/customs/prediction_request_rag/prediction_request_rag.py:74]: When running this tool, we get the following error:
```
ValueError: Model gpt-3.5-turbo-0125 not supported.
```

File: packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py [hunk 7]
```diff
@@ -0,0 +1,794 @@
...
+    "temperature": 0,
+}
+MAX_TOKENS = {
+    "gpt-3.5-turbo-0125": 16385,
+    "gpt-4-0125-preview": 8192,
+}
+ALLOWED_TOOLS = [
+    "prediction-request-reasoning",
+]
+TOOL_TO_ENGINE = {tool: "gpt-4-0125-preview" for tool in ALLOWED_TOOLS}
```
Comment (0xArdi) [inline packages/napthaai/customs/prediction_request_reasoning/prediction_request_reasoning.py:75]: Same issue here:
```
ValueError: Model gpt-4-0125-preview not supported.
```

**Review by 0xArdi — CHANGES_REQUESTED** [review]
There are issues with the new models introduced in this PR. They don't seem to be supported by either the OpenAI library we are using, or the OpenAI API Key we have.

**Comment by richardblythman** [comment]
I think this should fix the issue @0xArdi

**Review by 0xArdi — APPROVED** [review]

File: packages/packages.json [hunk 8]
```diff
@@ -23,13 +23,14 @@
...
+        "skill/valory/task_execution/0.1.0": "bafybeigt2yb6t7cglwqnpefsiuhzlxq6746e7m2bwm4niej23ptcduplqq",
         "skill/valory/websocket_client/0.1.0": "bafybeidwntmkk4b2ixq5454ycbkknclqx7a6vpn7aqpm2nw3duszqrxvta",
         "skill/valory/subscription_abci/0.1.0": "bafybeiclwa2u24nv4yzke6s24qxz2uxc6tpniyjxzprlb7oinkiijdhlgy",
-        "agent/valory/mech/0.1.0": "bafybeifrlwdaqfnychkmnpirix43zfbamgukkpziyuy7erze7owtbiuqc4",
-        "service/valory/mech/0.1.0": "bafybeifmcgdmynv7ww6x334ygt6eyljqepoddtaqhlywfi2jlqzhkecsu4"
+        "agent/valory/mech/0.1.0": "bafybeiblt6nix4zmfp35wcmyonkthp67mrauq3shqzvrb3423vmvp2j7vq",
+        "service/valory/mech/0.1.0": "bafybeia2pasegnizk6ozaheilsoduq3unwawmwfe572rj7e53ibvl42drm"
     },
     "third_party": {
+        "custom/napthaai/prediction_request_reasoning/0.1.0": "bafybeiaaefuqy47fv3iuh7qz3yd6ld6fkrt73ixdsilt3uyysr655h3vee",
```
Comment (jmoreira-valory) [inline packages/packages.json:33]: @0xArdi Why is this a third-party package?

**Review by dvilelaf — APPROVED** [review]

File: packages/packages.json (hunk 8 above)
Comment (0xArdi) [reply packages/packages.json]: it was on accident, has been addressed

**Review by jmoreira-valory — APPROVED** [review]

**Pull request merged**